File: 0015-3sum/0015-3sum.ts
function threeSum(nums: number[]): number[][] {
    nums = nums.sort((a,b) => a-b);
    let ans= []

    for (let i = 0; i < nums.length - 2; i++) {
        if (nums[i] === nums[i - 1]) continue;

        let curr = nums[i]
        let l = i+1;
        let r = nums.length-1
        while (l<r) {
            let sum = curr + nums[l] + nums[r]
            if (sum == 0) {
                ans.push([curr, nums[l], nums[r]])

                while (nums[l] === nums[l+1]) l++;
                while (nums[r] === nums[r-1]) r--;
            }

            if (sum > 0) {
                r--;
            } else {
                l++;
            }
        }
    }

    return ans
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        ans = []
        i = 0
        while i < len(nums) - 2:
            if i > 0 and nums[i] == nums[i-1]:
                i += 1
                continue 
            
            curr = nums[i]

            left = i+1
            right = len(nums)-1
            while left < right:
                val = curr + nums[left] + nums[right]

                if val == 0:
                    ans.append([curr, nums[left], nums[right]])

                    left += 1
                    right -= 1

                    while left < right and nums[left] == nums[left - 1]:
                        left += 1

                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                elif val < 0:
                    left += 1
                else:
                    right -= 1
            
            i += 1
            
        return ans